File: android/src/main/java/com/codenameakshay/async_wallpaper/AsyncWallpaperPlugin.java
package com.codenameakshay.async_wallpaper;

import android.app.WallpaperManager;
import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.provider.MediaStore;
import android.text.TextUtils;
import android.util.Log;
import android.util.Pair;
import androidx.annotation.RequiresApi;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Map;

import com.squareup.picasso.Picasso;
import com.squareup.picasso.Target;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodChannel;

import com.codenameakshay.async_wallpaper.PigeonApi;

public class AsyncWallpaperPlugin implements FlutterPlugin {
    private MethodChannel channel;
    private PigeonApiImpl pigeonApiImpl;

    @Override
    public void onAttachedToEngine(FlutterPlugin.FlutterPluginBinding binding) {
        pigeonApiImpl = new PigeonApiImpl(binding.getApplicationContext());
        PigeonApi.WallpaperApi.setUp(binding.getBinaryMessenger(), pigeonApiImpl);
    }

    @Override
    public void onDetachedFromEngine(FlutterPlugin.FlutterPluginBinding binding) {
        PigeonApi.WallpaperApi.setUp(binding.getBinaryMessenger(), null);
    }
}

File: android/src/main/kotlin/com/codenameakshay/async_wallpaper/AsyncWallpaperPlugin.kt
package com.codenameakshay.async_wallpaper

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding

/**
 * Owns one Pigeon implementation per Flutter engine and forwards only a currently attached
 * Activity to it. The implementation stores that Activity weakly, so configuration changes and
 * engine teardown cannot retain an old window.
 */
class AsyncWallpaperPlugin : FlutterPlugin, ActivityAware {
  private var implementation: PigeonApiImpl? = null

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    implementation?.shutdown()
    PigeonApiImpl(binding.applicationContext).also { api ->
      implementation = api
      WallpaperApi.setUp(binding.binaryMessenger, api)
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    WallpaperApi.setUp(binding.binaryMessenger, null)
    implementation?.shutdown()
    implementation = null
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    implementation?.attachActivity(binding.activity)
  }

  override fun onDetachedFromActivityForConfigChanges() {
    implementation?.detachActivity()
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    implementation?.attachActivity(binding.activity)
  }

  override fun onDetachedFromActivity() {
    implementation?.detachActivity()
  }
}
